### DISCUSSION TITLE
How to handle lifetimes within mock! block?

### DISCUSSION THREAD

**Discussion by mknet** [asomers/mockall#293] [Questions] [answered]
Hello everybody,

could you give me a hint on how to deal with lifetimes within the mock! macro block?

I want to do something like this:

```rust 
use mockall::*;
use mockall::predicate::*;
use std::fmt;

#[cfg(test)]
mod tests {
    use super::*;

    mock! {
        MyStruct {}

        impl fmt::Debug for MyStruct { 
            fn fmt(&self, f: &mut fmt::Formatter<'a>) -> fmt::Result {
                f.write_str("MyStruct")
            }
        }
    }

    #[test]
    fn test_clone() {
        let mut mock1 = MockMyStruct::new();
        mock1.expect_fmt()
            .returning(|f| f.write_str("MyMock"));

        assert_eq!("MyMock", format!("{:?}", mock1));
    }
}
```

and I get a result like:

```
error[E0261]: use of undeclared lifetime name `'a`
  --> src/kapitel_5_5_2_3_externe_traits.rs:13:50
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'a>) -> fmt::Result {
   |                                                  ^^ undeclared lifetime
...
17 |     }
   |      - lifetime `'a` is missing in item created through this procedural macro
   |
   = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
help: consider making the bound lifetime-generic with a new `'a` lifetime
   |
9  |     for<'a> mock! {
   |     ^^^^^^^
```

But the suggested `for<'a> mock` is not of course working. 

I'm looking forward to any hint.

Thanks and take care
Marcel

**Comment by asomers** [comment, marked as answer]
LOL, what timing!  The easiest course of action for you would be to upgrade Mockall to the latest version in git.  As of yesterday, it will automatically generate a `Debug` impl.
But if you don't want to do that, then write the function signature like this.  I think all three of these forms will work:
* `pub fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error>;`
* `pub fn fmt<'a>(&self, f: &mut Formatter<'a>) -> Result<(), Error>;`
* `pub fn fmt(&self, f: &mut Formatter -> Result<(), Error>;`

**Reply by mknet** [reply]
Of course! Well, I guess it was too late, yesterday.

I missed the missing lifetime after the function name...

`fn fmt<'a>(&self, f: &mut fmt::Formatter<'a>) -> fmt::Result` works for me

**Comment by asomers** [comment]
(also, you get to serve as Mockall's Discussions guinea pig.  Let's see if Discussions are a better way to ask for help than Issues).

**Reply by mknet** [reply]
Gotcha
